hnn_core/batch_simulate.py

```diff
@@ -0,0 +1,157 @@
+
+
+class BatchSimulate:
+    def __init__(self, set_params, net_name='jones', tstop=170,
+                 dt=0.025, n_trials=1):
+        """
+        Initialize the BatchSimulate class.
+
+        Parameters
+        ----------
+        set_params : func
+            User-defined function that sets parameters in network drives.
+            `set_params(net, params) -> None`
+        net_name : str
+            The name of the network model to use. Default is `jones`.
+        tstop : float, optional
+            The stop time for the simulation. Default is 170 ms.
+        dt : float, optional
+            The time step for the simulation. Default is 0.025 ms.
+        n_trials : int, optional
+            The number of trials for the simulation. Default is 1.
+        """
+        self.set_params = set_params
+        self.net_name = net_name
+        self.tstop = tstop
+        self.dt = dt
+        self.n_trials = n_trials
+
+    def run(self, param_grid, return_output=True, combinations=True, n_jobs=1):
+        """
+        Run batch simulations.
+
+        Parameters
+        ----------
+        param_grid : dict
+            Dictionary with parameter names and ranges.
+        return_output : bool, optional
+            Whether to return the simulation outputs. Default is True.
+        combinations : bool, optional
+            Whether to generate the Cartesian product of the parameter ranges.
+            If False, generate combinations based on corresponding indices.
+            Default is True.
+        n_jobs : int, optional
+            Number of parallel jobs. Default is 1.
+
+        Returns
+        -------
+        results : list
+            List of simulation results if return_output is True.
+        """
+        param_combinations = self._generate_param_combinations(
+            param_grid, combinations)
+        results = self.simulate_batch(param_combinations, n_jobs=n_jobs)
+
+        if return_output:
+            return results
+
+    def simulate_batch(self, param_combinations, n_jobs=1):
+        """
+        Simulate a batch of parameter sets in parallel.
+
+        Parameters
+        ----------
+        param_combinations : list
+            List of parameter combinations.
+        n_jobs : int, optional
+            Number of parallel jobs. Default is 1.
+
+        Returns
+        -------
+        res: list
+            List of dictionaries containing simulation results.
+            Each dictionary contains the following keys:
+            - `net`: The network model used for the simulation.
+            - `dpl`: The simulated dipole.
+            - `param_values`: The parameter values used for the simulation.
+        """
+        res = Parallel(n_jobs=n_jobs, verbose=50)(
+            delayed(self._run_single_sim)(
+                params) for params in param_combinations)
+        return res
+
+    def _run_single_sim(self, param_values):
```

Comment:
In the next PR let's add an option to pass a callable in the `__init__()` that overrides this function